File: 02-mobile-client/mobile-frontend/lib/models/promo.dart
class Promo {
  final String id;
  final String title;
  final String description;
  final String promoType;
  final double discountValue;
  final double minOrderAmount;
  final DateTime startDate;
  final DateTime endDate;
  final String? imageUrl;
  final String? imageId;
  final String status;
  final bool isActive;

  Promo({
    required this.id,
    required this.title,
    required this.description,
    required this.promoType,
    required this.discountValue,
    required this.minOrderAmount,
    required this.startDate,
    required this.endDate,
    this.imageUrl,
    this.imageId,
    required this.status,
    required this.isActive,
  });

  factory Promo.fromJson(Map<String, dynamic> json) {
    return Promo(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      promoType: json['promoType'] ?? '',
      discountValue: (json['discountValue'] ?? 0).toDouble(),
      minOrderAmount: (json['minOrderAmount'] ?? 0).toDouble(),
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      imageUrl: json['imageUrl'],
      imageId: json['imageId'],
      status: json['status'] ?? '',
      isActive: json['isActive'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'promoType': promoType,
      'discountValue': discountValue,
      'minOrderAmount': minOrderAmount,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'imageUrl': imageUrl,
      'imageId': imageId,
      'status': status,
      'isActive': isActive,
    };
  }

  Promo copyWith({
    String? id,
    String? title,
    String? description,
    String? promoType,
    double? discountValue,
    double? minOrderAmount,
    DateTime? startDate,
    DateTime? endDate,
    String? imageUrl,
    String? imageId,
    String? status,
    bool? isActive,
  }) {
    return Promo(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      promoType: promoType ?? this.promoType,
      discountValue: discountValue ?? this.discountValue,
      minOrderAmount: minOrderAmount ?? this.minOrderAmount,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      imageUrl: imageUrl ?? this.imageUrl,
      imageId: imageId ?? this.imageId,
      status: status ?? this.status,
      isActive: isActive ?? this.isActive,
    );
  }

  /// Check if the promo is currently active
  bool get isCurrentlyActive {
    final now = DateTime.now();
    return isActive && now.isAfter(startDate) && now.isBefore(endDate);
  }

  /// Check if the promo has an image (either GridFS or URL)
  bool get hasImage => (imageId != null && imageId!.isNotEmpty) || 
                      (imageUrl != null && imageUrl!.isNotEmpty);

  /// Get formatted discount text
  String get discountText {
    switch (promoType) {
      case 'percentage':
        return '${discountValue.toInt()}% OFF';
      case 'fixed':
        return '₱${discountValue.toInt()} OFF';
      case 'buy_one_get_one':
        return 'Buy 1 Get 1';
      default:
        return 'Special Offer';
    }
  }
}

File: 01-web-application/backend/models/Promo.js
const mongoose = require('mongoose');

const promoSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
    trim: true
  },
  description: {
    type: String,
    required: true,
    trim: true
  },
  promoType: {
    type: String,
    required: true,
    enum: ['Percentage Discount', 'Fixed Amount Discount', 'Buy One Get One', 'Free Item', 'Loyalty Points Bonus']
  },
  discountValue: {
    type: Number,
    required: true,
    min: 0
  },
  minOrderAmount: {
    type: Number,
    default: 0,
    min: 0
  },
  startDate: {
    type: Date,
    required: true
  },
  endDate: {
    type: Date,
    required: true
  },
  usageLimit: {
    type: Number,
    default: null, // null means unlimited
    min: 1
  },
  status: {
    type: String,
    required: true,
    enum: ['Active', 'Inactive', 'Scheduled', 'Expired'],
    default: 'Active'
  },
  imageUrl: {
    type: String,
    default: null
  },
  isActive: {
    type: Boolean,
    default: true
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Admin',
    required: true
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Admin'
  }
}, {
  timestamps: true
});

// Index for efficient queries
promoSchema.index({ status: 1, isActive: 1, startDate: 1, endDate: 1 });
promoSchema.index({ createdAt: -1 });

// Virtual for checking if promo is currently active based on dates
promoSchema.virtual('isCurrentlyActive').get(function() {
  const now = new Date();
  return this.isActive && 
         this.status === 'Active' && 
         this.startDate <= now && 
         this.endDate >= now;
});

// Method to update status based on dates
promoSchema.methods.updateStatus = function() {
  const now = new Date();
  
  if (!this.isActive) {
    this.status = 'Inactive';
  } else if (this.endDate < now) {
    this.status = 'Expired';
  } else if (this.startDate > now) {
    this.status = 'Scheduled';
  } else {
    this.status = 'Active';
  }
  
  return this.save();
};

// Static method to get active promos for public display
promoSchema.statics.getActivePromos = function() {
  const now = new Date();
  return this.find({
    isActive: true,
    status: 'Active',
    startDate: { $lte: now },
    endDate: { $gte: now }
  }).sort({ createdAt: -1 });
};

module.exports = mongoose.model('Promo', promoSchema);
